CDCgov/prime-reportstream
# Consider topics when routing in Universal Pipeline

This PR adjusts the UP routing code to only consider receivers of the same topic as the sender. The change to the router was _very_ small, so the bulk of the changeset is the adjustments to the RoutingTests.kt file.

Test Steps:

- Confirm that etor sender can send to etor receiver

  1. Reload updated etor settings with this command `./prime multiple-settings set --input ./settings/staging/0149-etor.yml`
  2. Send a lab order report using this postman collection: 
[Temp Collection.postman_collection_json.txt](https://github.com/CDCgov/prime-reportstream/files/11548178/Temp.Collection.postman_collection_json.txt)
  3. Query the submission history to verify that the report is sent to `simulated-lab`, the topic is properly recorded, and check the sftp folder to look for the resulting HL7V2 file.
- Confirm that etor sender cannot send to full-elr receiver
  1. Update  `./settings/staging/0149-etor.yml` to set `simulated-lab`'s topic to `full-elr`.
  2. Reload updated etor settings with this command `./prime multiple-settings set --input ./settings/staging/0149-etor.yml`
  3. Send a lab order report using this postman collection: 
[Temp Collection.postman_collection_json.txt](https://github.com/CDCgov/prime-reportstream/files/11548178/Temp.Collection.postman_collection_json.txt)
  4. Query the submission history to verify that the report was not delivered and the topic is properly recorded.
- Confirm that full-elr sender can send to full-elr receiver
  1. Update  `./settings/staging/0149-etor.yml` to set `etor-service-sender`'s topic to `full-elr`.
  2. Reload updated etor settings with this command `./prime multiple-settings set --input ./settings/staging/0149-etor.yml`
  3. Send a lab order report using this postman collection: 
[Temp Collection.postman_collection_json.txt](https://github.com/CDCgov/prime-reportstream/files/11548178/Temp.Collection.postman_collection_json.txt)
  4. Query the submission history to verify that the report is sent to `simulated-lab`, the topic is properly recorded, and check the sftp folder to look for the resulting HL7V2 file.
- Confirm that full-elr sender cannot send to etor receiver
  1. Update  `./settings/staging/0149-etor.yml` to set `simulated-lab`'s topic to `etor-ti`.
  2. Reload updated etor settings with this command `./prime multiple-settings set --input ./settings/staging/0149-etor.yml`
  3. Send a lab order report using this postman collection: 
[Temp Collection.postman_collection_json.txt](https://github.com/CDCgov/prime-reportstream/files/11548178/Temp.Collection.postman_collection_json.txt)
  4. Query the submission history to verify that the report was not delivered and the topic is properly recorded.



## Changes
- Changed the lambda on the initial retrieval of receivers during `applyFilters` to only include receivers with matching topic to a new input parameter
- Updated RoutingTests.kt to use a helper function and reduce code duplication
- Added tests to cover the 4 cases listed in the "Test Steps" section above

## Checklist

### Testing

## Linked Issues
- Addresses #9495
